### DISCUSSION TITLE
[v5] BelongsTo, preloading and associate() question

### DISCUSSION THREAD

**Discussion by 42pe** [adonisjs/core#2224] [Help] [answered]
So... I'm going to try to explain this, bear with me, I'm new at Adonis.

I have a **ProjectOwner** Model (basically a "Company" so to speak), where I have a relationship to **User** as who owns it:
```
@belongsTo(() => User, { localKey: 'id', foreignKey: 'user_id' })
  public owner: BelongsTo<typeof User>
```

## Question: Association

When doing this: 

```
const po = await ProjectOwner.create(data)
      po.related('owner').associate(user)
```

I couldn't get it to associate a new ProjectOwner without defining the `user_id` column in the model, besides the relationship, as such:

```
  @column()
  protected user_id: number
```

Without that I got the following error:
`E_MISSING_MODEL_ATTRIBUTE: "ProjectOwner.owner" expects "user_id" to exist on "ProjectOwner" model, but is missing`

So, **¿Is it correct that I need to add the column `used_id` besides defining the relationship?**

**Comment by thetutlage** [comment, marked as answer]
Yes, the model needs the foreignKeys to be available on the model itself. Infact, the `foreignKey` property inside relationship options is the name of the property on the model and not the table.

To summarize it, Models form relationships with each other and not the database tables.

**Reply by 42pe** [reply]
Clear as water, thank you so much!

**Comment by jonathanestefani** [comment]
Is it possible to make a composite key?

**Reply by thetutlage** [reply]
Nope. Composite keys are not possible as of now. It requires a lot of changes acorss the entire codebase and I am not planning to invest time for that right now.

**Comment by jonathanestefani** [comment]
An example:

@belongsTo(() => Invoices, { localKey: ['company_id, 'branch', 'store'], foreignKey: ['company_id, 'branch', 'store'] })
  public orders: BelongsTo<typeof Order>

**Comment by tuxaanand** [comment]
I've just started with AdonisJS. Thank you all for such a beautiful framework.

In KeysExtractor, I tried adding the column dynamically if the attribute doesn't exist already as follows and it worked. 

Would it be possible to incorporate this behavior into Lucid? Will this impact anything else?

```

      let attribute = model.$getColumn(key)

      if (!attribute) {
        model.$addColumn(key, {})
        attribute = model.$getColumn(key)
        if (!attribute) {
          throw new Exception(
            `"${relationRef}" expects "${key}" to exist on "${model.name}" model, but is missing`,
            500,
            'E_MISSING_MODEL_ATTRIBUTE'
          )
        }
      }


```
